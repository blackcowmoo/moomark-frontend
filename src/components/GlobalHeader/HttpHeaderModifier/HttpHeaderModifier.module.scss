.container {
  width: 3rem;
}

.contents{
  min-width: 50rem;
  padding: 1rem;

}