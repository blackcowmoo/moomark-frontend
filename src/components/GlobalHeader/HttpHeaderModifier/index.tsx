import Modal from '@components/common/Modal';
import styles from './HttpHeaderModifier.module.scss';
import { useModal } from 'utils/hooks/useModal';
const index = () => {
  const { isShown, toggle } = useModal();
  return (
    <div>
      <button className={styles.button} onClick={toggle}>
        🐮
      </button>
      <Modal title='Http Client Header Custom' isShown={isShown} onClose={toggle} content={<>custom contents</>} />
    </div>
  );
};

export default index;
