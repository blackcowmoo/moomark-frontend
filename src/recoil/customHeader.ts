import { atom, selector } from 'recoil';

export const defaultHeaderOption = { 'x-moom-route': '' };

export const customHeaderAtom = atom<Object>({
  key: 'customheader',
  default: defaultHeaderOption,
});

export const customHeaderState = selector({
  key: 'customHeader',
  get: ({ get }) => {
    const customHeader = get(customHeaderAtom);
    return customHeader;
  },
});
