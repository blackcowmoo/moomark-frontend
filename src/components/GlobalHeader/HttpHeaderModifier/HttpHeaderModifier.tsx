import { useEffect, useState } from 'react';
import { useRecoilState } from 'recoil';
import { customHeaderAtom, defaultHeaderOption } from 'recoil/customHeader';
import styles from './HttpHeaderModifier.module.scss';

interface ICustomOption {
  customKey: string;
  customValue: string;
}

const HttpHeaderModifier = () => {
  const [customOption, setCustomOption] = useState<ICustomOption>({ customKey: '', customValue: '' });
  const { customKey, customValue } = customOption;
  const [customHeader, setCustomHeader] = useRecoilState(customHeaderAtom);
  const localStorageKey = 'mooMark-dev-customHeader';

  useEffect(() => {
    const savedCustomHeader = localStorage.getItem(localStorageKey);
    if (savedCustomHeader) {
      setCustomHeader(JSON.parse(savedCustomHeader));
    }
  }, []);

  useEffect(() => {
    localStorage.setItem(localStorageKey, JSON.stringify(customHeader));
  }, [customHeader]);

  const clone = (obj: any) => ({ ...obj });

  const changeHeaderValue = (key: string, value: string) => {
    return { ...customHeader, [key]: value };
  };

  const onChangeHeaderValue = (key: string, e: React.ChangeEvent<HTMLInputElement>) => {
    setCustomHeader(changeHeaderValue(key, e.target.value));
  };

  const removeOption = (object: Object, key: string) => {
    const clonedObj = clone(object);
    delete clonedObj[key];

    setCustomHeader(clonedObj);
  };

  const clearCustomOption = () => {
    setCustomOption({ customKey: '', customValue: '' });
  };

  const resetHeaderOption = () => {
    setCustomHeader(defaultHeaderOption);
    clearCustomOption();
  };

  const addCustomOption = (key: string, value: string) => {
    setCustomHeader((prev) => {
      return { ...prev, [key]: value };
    });
    clearCustomOption();
  };

  const handleCustomKey = (e: React.ChangeEvent<HTMLInputElement>) => {
    setCustomOption((prev) => {
      return { ...prev, customKey: e.target.value };
    });
  };

  const handleCustomValue = (e: React.ChangeEvent<HTMLInputElement>) => {
    setCustomOption((prev) => {
      return { ...prev, customValue: e.target.value };
    });
  };

  return (
    <div className={styles.contents}>
      {Object.entries(customHeader).map(([key, value]) => (
        <div className={styles.item} key={key}>
          <input value={key} />
          <input onChange={(e) => onChangeHeaderValue(key, e)} value={value} />
          <button onClick={() => removeOption(customHeader, key)}>remove</button>
        </div>
      ))}
      <div className={styles.inputContainer}>
        <input value={customKey} onChange={(e) => handleCustomKey(e)} />
        <input value={customValue} onChange={(e) => handleCustomValue(e)} />
        <button onClick={() => addCustomOption(customKey, customValue)}>add</button>
      </div>
      <button className={styles.resetButton} onClick={() => resetHeaderOption()}>
        Reset Option
      </button>
      <p>check value in localStorage Key: {localStorageKey}</p>
      <div className={styles.result}>{JSON.stringify(customHeader)}</div>
    </div>
  );
};

export default HttpHeaderModifier;
