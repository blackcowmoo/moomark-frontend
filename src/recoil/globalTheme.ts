import { atom } from 'recoil';

export type Theme = 'light' | 'dark' | null;

export const globalThemeState = atom<Theme>({
  key: 'globalThemeState',
  default: null,
});
