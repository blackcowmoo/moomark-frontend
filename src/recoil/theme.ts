import { atom } from 'recoil';

export type Theme = 'light' | 'dark' | null;

export const themeState = atom<Theme>({
  key: 'globalThemeState',
  default: null,
});
