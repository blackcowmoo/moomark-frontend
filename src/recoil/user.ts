import { atom } from 'recoil';

export interface IUser {
  nickname: string;
  email: string;
  picture: null | string;
  role: string | null;
}

export const initialUserState: IUser = { nickname: '', email: '', picture: null, role: null };

export const userRecoilState = atom<IUser>({
  key: 'currentUser',
  default: initialUserState,
});
