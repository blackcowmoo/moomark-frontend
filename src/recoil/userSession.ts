import { atom, selector } from 'recoil';

interface IUserSession {
  id: null | string;
  userName: null | string;
}

export const userSessionAtom = atom<IUserSession>({
  key: 'userSession',
  default: { id: null, userName: null },
});

export const loginUserState = selector({
  key: 'loginUserState',
  get: ({ get }) => {
    const userSession = get(userSessionAtom);
    const { userName } = userSession;

    return { userName };
  },
});
