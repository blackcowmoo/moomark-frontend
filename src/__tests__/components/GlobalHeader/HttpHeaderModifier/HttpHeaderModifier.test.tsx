import { mount } from 'enzyme';
import { RecoilRoot } from 'recoil';
import HttpHeaderModifier from '@components/GlobalHeader/HttpHeaderModifier/HttpHeaderModifier';

describe('<HttpHeaderModifier/>', () => {
  it('matches snapshot', () => {
    const wrapper = mount(
      <RecoilRoot>
        <HttpHeaderModifier />
      </RecoilRoot>,
    );
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
