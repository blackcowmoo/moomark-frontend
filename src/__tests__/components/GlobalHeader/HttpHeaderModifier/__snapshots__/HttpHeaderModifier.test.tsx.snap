// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HttpHeaderModifier/> matches snapshot 1`] = `
"<RecoilRoot>
  <RecoilRoot_INTERNAL>
    <Batcher setNotifyBatcherOfChange={[Function (anonymous)]} />
    <HttpHeaderModifier>
      <div className=\\"wrapper\\">
        <div className=\\"headerWrapper\\">
          <button onClick={[Function: onSaveInputArray]}>
            save
          </button>
          <button onClick={[Function: onClick]}>
            reset input
          </button>
        </div>
        <div className=\\"inputContainer\\">
          <div>
            input new option
          </div>
          <input value=\\"\\" onChange={[Function: onChange]} />
          <input value=\\"\\" onChange={[Function: onChange]} />
          <button onClick={[Function: onClick]}>
            add
          </button>
        </div>
        <div className=\\"textareaWrapper\\">
          <div>
            Result
          </div>
          <textarea className=\\"result\\" value=\\"{\\"x-moom-route\\":\\"\\"}\\" onChange={[Function: onChangeTextarea]} />
          <button onClick={[Function: onSaveTextarea]}>
            textarea save
          </button>
          <button className=\\"resetButton\\" onClick={[Function: resetHeaderOption]}>
            reset header
          </button>
        </div>
      </div>
    </HttpHeaderModifier>
  </RecoilRoot_INTERNAL>
</RecoilRoot>"
`;
